SIP minutes: 2019 December (#1614)
scala/docs.scala-lang: _sips/minutes/2019-12-18-sip-minutes.md

* Adriaan Moors ([@adriaanm](https://github.com/adriaanm)), Lightbend
* Seth Tisue ([@SethTisue](https://github.com/SethTisue)), Lightbend
* Iulian Dragos ([@dragos](https://github.com/dragos)), Triplequote
* Lukas Rytz ([@lrytz](https://twitter.com/lrytz)), visiting from Lightbend
* Dale Wijnand ([@dwijnand](https://twitter.com/dwijnand)), secretary

## Not present

* Miles Sabin ([@milessabin](https://github.com/milessabin)), Independent
* Heather Miller ([@heathermiller](https://github.com/heathermiller)), CMU
* Josh Suereth ([@jsuereth](https://github.com/jsuereth)), Independent

## Proceedings

### Organisational changes

Darja gave an update on some recently agreed upon changes to help streamline the SIP process for Scala 3.  Each
member of the committee will "own" a certain number of features, which they will either champion to the
committee and/or which they will spend time investigating and trying to poke holes into.  The goals are:

1. make sure that a good number of people exercise the feature, try to break it, refine it and raise any red
   flags about it;
2. provide a detailed review of the feature, especially for the contentious ones; and, finally

> 3. straw poll will be take on each feature, as a smoke test for the real vote

Reviewer: ```suggestion
3. straw poll will be taken on each feature, as a smoke test for the real vote
```

Sébastien is reminded that he hasn't yet talked about `.nn`.  `nn` is an extension method added to `A | Null`
that, when invoked, is an assertiong that a value is not null (thus "nn") and will throw a NPE if it is and
return `A` if it's not.

Sébastien shares he doesn't like that `.nn` is available everywhere, without even an import.  Guillaume says
it's a bit like `.get` on Option.  Martin says he would defend .get and .nn, say that users should not use them
if they don't like them.  Martin also says that maybe after the Scala 3 transition we could deprecate it and/or
move it, but for now he thinks it's good to have universally available.

Seth asks Martin: is null only for performance-critical code, or is that just one of the use cases?
Martin: Java interop or performance-critical code
Seth: Doesn't the UncheckedNull cover the Java case?
Martin: well, no, you might want to use A | Null in some code adjacent to some Java interop code, which wouldn't
be terribly low level code.  So in that case you might continue to use null.
Seb: an easier way to answer is that UncheckedNull is only to select a member.  You cannot assign a String |

> UncheckedNull to a String.  If you want to string the value in a String, it will be checked then, at runtime.

Reviewer: ```suggestion
UncheckedNull to a String.  If you want to store the value in a String, it will be checked then, at runtime.
```

still finds that UncheckedNull is a very easy way to turn a blind eye on nulls and fears its existing will
continue to allow lots of NPEs to be generated.  He thinks it's very easy not to realise that these are
unchecked nulls, using a Java method, and you don't get any warnings.  Martin says how no, as soon as you
typecheck it in Scala as String, then that means String, not String | UncheckedNull, so that's theoretically
your firewall.

Guillaume also shares how with the proposal also come a bunch of annotations for the Java standard library that
are supposed to know which methods take or return null, so for thins that return null Scala could make it type
as `A | Null` instead of `A | UncheckedNull`.

Sébastien shares that some users don't like the magic of UncheckedNull, so maybe that could be behind a language
flag, though there are language dialect concerns with language flags.

Seth then asks about the status of the -Y flag associated with this feature: is the intention of the feature to
be on by default?  Sébastien says: hopefully, if it all works out.  Martin says maybe it could be turned on by

> by Scala 3.1.  Guillaume says he's confused what it means for some libraries to enable the flag and some not to.

Reviewer: ```suggestion
default by Scala 3.1.  Guillaume says he's confused what it means for some libraries to enable the flag and some not to.
```

> Martin summarises that, basically, we ened to get to the poin where we have it on by default.

Reviewer: ```suggestion
Martin summarises that, basically, we need to get to the point where we have it on by default.
```

Sébastien, looking at the YouTube chat, shares Eugene Yokota's question: "what about `var x: String = _`?"
Sébastien shares how it's "evil" and how it should've been removed a long time ago.  Martin counters that it's

> quite the opposite: it should be taken to mean that it's assigning null, it should be taken to meant that the

Reviewer: ```suggestion
quite the opposite: it should not be taken to mean that it's assigning null, it should be taken to mean that the
```

assigning null.

Dale states how it seems there aren't any good ergonomics for using a value of type `A | Null`.  Sébastien says
that's a good thing: use if/else or Option.  Dale clarifies that he was trying to make a case for promoting
`UOption` (from [sjrd/scala-unboxed-option](https://github.com/sjrd/scala-unboxed-option)).  Sébastien says that
the meeting is running out of time, but he has a solution with UOption.  Martin seems pleased and ask "Can we
get it, please, for 3.0?".  Seth comments how Sébastien comment sounds like "too long to fit in this margin". :D

### Review the "main functions" SIP

Docs: <https://dotty.epfl.ch/docs/reference/changed-features/main-functions.html>

Sébastien shares that there are only a few minutes left in the meeting but the gist is that DelayedInit is gone
(so far) and the obvious question is: what do you do for a simple main method?  The idea for now is to have an
`@main` annotation and you can write a `def foo`, which takes parameters, and you write the body inside, and,
since we have top level functions, you can write that at the top level, which is even better and everyone is
happy.  But there are details, of course.

> Maritn states it lacks a champion and asks if anyone wants to champion it.  Seth agrees and, so, Sébastien asks

Reviewer: ```suggestion
Martin states it lacks a champion and asks if anyone wants to champion it.  Seth agrees and, so, Sébastien asks
```

him to open a contributors thread on it.

## Next

The next meeting will be on the last week of January, at 17:00 CET.  The same will happen in February.  In March

> the Committee will meeting for a 3 day retreat but it will still come online for an hour, to provide a summary

Reviewer: ```suggestion
the Committee will meet for a 3 day retreat but it will still come online for an hour, to provide a summary
```
